WIP: list-mechanisms: initial implementation

t184256 opened the pull request:
Resolves: #563

To be addressed:
* is mixing APIs OK?
* which printing API should I use?
* there are no tests

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,303 @@
+		goto cleanup;
+	}
+
+	p11_list_printer_init (&printer, stdout, 0);
```
There are some duplicated lines between line 139 to 187 (`p11_kit_modules_load_and_initialize (0)` is called twice, etc.). Maybe a copy and paste error?

t184256 replied:
indeed, removed

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,303 @@
+	CK_SLOT_ID slot = 0;
+	CK_MECHANISM_INFO mechanism_info;
+	CK_MECHANISM_TYPE_PTR mechanisms = NULL;
+	CK_MECHANISM_TYPE mechanism = CKP_INVALID_ID;
```
Type mismatch: `CKP_INVALID_ID` is defined as CK_PROFILE_ID

t184256 replied:
side-stepped

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,303 @@
+			goto cleanup;
+		}
+
+		mechanisms = malloc(sizeof(CK_MECHANISM_TYPE) * mechanisms_count);
```
```suggestion
		mechanisms = reallocarray (NULL, mechanisms_count, sizeof (CK_MECHANISM_TYPE));
```
Maybe it would also be possible to reuse the previously allocated `mechanisms` by resizing.

t184256 replied:
right, and it was also leaking. used reallocarray

t184256 commented:
addressed feedback, added a test against mock data, added printing token labels

I don't see how reasonably use the printing API for what I'm doing, so I just hope mixing it with printfs is OK.

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,288 @@
+		}
+
+		slot = p11_kit_iter_get_slot (iter);
+		if (slot == 0) {
```
0 is a valid slot ID; we can safely assume that slot is always valid.

t184256 replied:
removed the check

ueno commented on p11-kit/test-list-mechanisms.sh:
```diff
@@ -0,0 +1,47 @@
+0x80000002 (unknown): sign verify key-size=2048-2048
+EOF
+
+	if ! "$abs_top_builddir"/p11-kit/p11-kit-testable list-mechanisms -q pkcs11: > list.out; then
```
This URL matches all tokens? I would rather limit it to the first match, as `p11tool --list-mechanism` does.

t184256 replied:
> This URL matches all tokens?

Yes.

> I would rather limit it to the first match, as p11tool --list-mechanism does.

You mean, limit `list-mechanisms` subcommand to inspecting the first matching token? I'd disagree on that. First, I think it might introduce confusion in a scenario where the user accidentally matches several tokens. Second, AFAICS this behaviour won't be in line with the existing `list-*` subcommands that iterate over all matches.

ueno replied:
> AFAICS this behaviour won't be in line with the existing list-* subcommands that iterate over all matches

Not really:
- `p11-kit list-modules`: doesn't take URL
- `p11-kit list-profiles`: takes a token URL, uses first match only, because profiles are capabilities of a particular token
- `p11-kit list-objects`: takes an *object* URL, returns all matches

t184256 replied:
Changed to stop after first match. Testing 3 scenarios now: "pkcs11:", specific existing and specific non-existing token. Maybe when we have list-tokens we should also test the iteration you're suggesting.

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,288 @@
+
+	iter = p11_kit_iter_new (uri, P11_KIT_ITER_WITH_LOGIN | P11_KIT_ITER_WITH_TOKENS | P11_KIT_ITER_WITHOUT_OBJECTS);
+	if (iter == NULL) {
+		p11_message (_("failed to initialize iterator"));
```
As this is an internal error, it might make sense to use `p11_debug` instead.

t184256 replied:
changed to `p11_debug`

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,285 @@
+#include "config.h"
+
+#include "constants.h"
+#define P11_DEBUG_FLAG P11_DEBUG_LIB
```
```suggestion
#define P11_DEBUG_FLAG P11_DEBUG_TOOL
```

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,285 @@
+		goto cleanup;
+	}
+
+	if (p11_kit_uri_parse (token_str, P11_KIT_URI_FOR_OBJECT_ON_TOKEN, uri) != P11_KIT_URI_OK) {
```
We iterate over tokens, not objects on them, right?
```suggestion
	if (p11_kit_uri_parse (token_str, P11_KIT_URI_FOR_TOKEN, uri) != P11_KIT_URI_OK) {
```

ueno commented on p11-kit/list-mechanisms.c:
```diff
@@ -0,0 +1,281 @@
+
+	p11_list_printer_init (&printer, stdout, 0);
+	p11_kit_iter_begin (iter, modules);
+	rv = p11_kit_iter_next (iter);  // we will stop at the first token
```
```suggestion
	rv = p11_kit_iter_next (iter);
	if (rv != CKR_OK) {
		if (rv == CKR_CANCEL)
			p11_message (_("no matching token"));
		else
			p11_message (_("failed to find token: %s"), p11_kit_strerror (rv));
		goto cleanup;
	}
```

t184256 replied:
incorporated

ueno reviewed (approved):
Thank you, looks good to me!
